caproto/server/conversion.py

```diff
@@ -230,182 +231,433 @@ def group_to_device(group):
+
+    if record_type != 'base':
+        for mixin, mixin_info in MIXIN_SPECS.items():
+            has_fields = all(field in field_info
+                             for field in mixin_info)
+            types_match = all(
+                DBD_TYPE_INFO[field_info[field]['type']].data_type == mixin_info[field]
+                for field in mixin_info
+                if field in field_info
+            )
+            if has_fields and types_match:
+                # Add the mixin
+                result['mixin'] = [mixin.__name__]
+                # And remove those attributes from the subclass
+                for field in mixin_info:
+                    fields_by_attr.pop(field_to_attr[field])
+
+    for field_attr, field in fields_by_attr.items():
+        result['fields'].append(field)
+        try:
+            channeldata_attr = LINKABLE[field_attr]
+        except KeyError:
+            ...
         else:
-            yield from lines
-
-    linkable = {
-        'display_precision': 'precision',
-        'hihi_alarm_limit': 'upper_alarm_limit',
-        'high_alarm_limit': 'upper_warning_limit',
-        'low_alarm_limit': 'lower_warning_limit',
-        'lolo_alarm_limit': 'lower_alarm_limit',
-        'high_operating_range': 'upper_ctrl_limit',
-        'low_operating_range': 'lower_ctrl_limit',
-        # 'alarm_deadband': '',
-        'archive_deadband': 'log_atol',
-        'monitor_deadband': 'value_atol',
-    }
+            link = dict(field_attr=field_attr,
+                        channeldata_attr=channeldata_attr,
+                        kwargs=LINK_KWARGS.get(field_attr, {}),
+                        )
+            result['links'].append(link)
+
+    result['field_to_attr'] = field_to_attr
+    return result
+
+
+def _get_sort_id(name, list_):
+    'Returns tuple of (index in list, name)'
+    key1 = (list_.index(name)
+            if name in list_
+            else len(list_) + 1
+            )
+    return (key1, name)
+
+
+def generate_all_records_jinja(dbd_file, *, jinja_env=None,
+                               template='records.jinja2'):
+    try:
+        from caproto.tests.dbd import DbdFile
+        import jinja2
+    except ImportError as ex:
+        raise ImportError(f'An optional/testing dependency is missing: {ex}')
+
+    if jinja_env is None:
+        jinja_env = jinja2.Environment(
+            loader=jinja2.PackageLoader("caproto", "server"),
+            trim_blocks=True,
+            lstrip_blocks=True,
+        )
+
+    dbd_file = DbdFile.parse_file(dbd_file)
+    existing_record_list = ['base'] + list(records_mod.records)
+    records = {}
+    for record in ('base', ) + tuple(sorted(dbd_file.field_metadata)):
+        d = record_to_template_dict(record, dbd_file.field_metadata)
+        d['sort_id'] = _get_sort_id(record, existing_record_list)
+        records[record] = d
+
+    record_template = jinja_env.get_template(template)
+    return record_template.render(records=records)
+
+
+MENU_FIXES = {
```

Comment:
Same here...I think it's clear what this is doing but a comment would help.

Reply:
Resolved in 7b519675